### ISSUE TITLE
Is there a way to have multiple custom functions with different namespaces?

### ISSUE THREAD

**Issue by wh1t3cAt1k** [OfficeDev/Excel-Custom-Functions#206]
As I understand it would require setting up several extension points with type CustomFunctions, each of which would have a separate namespace.

Is it supported?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 94ad7803-d4f7-a700-7c7f-204cd8f34e4a
* Version Independent ID: 9c485bae-cab1-bdad-4c8e-baad2b12b22d
* Content: [Create custom functions in Excel - Office Add-ins](https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-overview)
* Content Source: [docs/excel/custom-functions-overview.md](https://github.com/OfficeDev/office-js-docs-pr/blob/master/docs/excel/custom-functions-overview.md)
* Product: **excel**
* Technology: **add-ins**
* GitHub Login: @o365devx
* Microsoft Alias: **o365devx**

**Comment by mscharlock** [comment]
@RuoyingLiang - Do you mind answering this one? I think this is possible but I want to check.

**Comment by RuoyingLiang** [comment]
@wh1t3cAt1k Custom function in one add-in don't support different namespace. Could you share with us your scenario to help us understand the requirement?

**Comment by mscharlock** [comment]
@RuoyingLiang - I will go ahead and make sure this is accurate in the docs! :)

**Comment by wh1t3cAt1k** [comment]
> @wh1t3cAt1k Custom function in one add-in don't support different namespace. Could you share with us your scenario to help us understand the requirement?

@RuoyingLiang the scenario is rather simple. We connect to back ends of different providers that expose their public APIs to us, but with similar functionalities, e.g. CalculateInterest().

We need to disambiguate clearly between providers in the naming. Currently we are only able this through prefixing:

CONTOSO.FOOCALCULATEINTEREST()
CONTOSO.BARCALCULATEINTEREST() 

etc.

Would be a bit nicer to say CONTOSOFOO.CALCULATEINTEREST() and CONTOSOBAR.CALCULATEINTEREST(). 

Basically that is our scenario.

**Comment by wh1t3cAt1k** [comment]
Nested namespaces like CONTOSO.FOO.CALCULATEINTEREST() would rock even more in terms of readability and intuitiveness.

**Comment by akrantz** [comment]
The name of the custom function can have a prefix, so instead of the function being CALCULATEINTEREST(), the actual function name can have a period and be FOO.CALCULATEINTEREST() and BAR.CALCULATEINTEREST().  If you have the namespace as "CONTOSO", then these would appear as CONTOSO.FOO.CALCULATEINTEREST() and CONTOSO.BAR.CALCULATEINTEREST().

**Comment by wh1t3cAt1k** [comment]
Thank you. This works for us.

@akrantz will this require calling CustomFunctions.associate? I trust JS functions cannot contain dotted identifiers.

Please @mscharlock clarify this in the documentation - maybe including the workaround suggested by @akrantz?

Close the issue yourself when ready please, and thank you again for helping out!

**Comment by akrantz** [comment]
The JavaScript function name cannot contain the period, but the name in the metadata certainly can. These names do not have to be the same.

I'd like to call out that the Yo Office template for custom functions has a webpack plugin that can generate the custom function metadata from the JavaScript/TypeScript source. It will generate the metadata name from the JavaScript function name but you can override this in the @customfunction tag. We made sure to allow for periods in this name for just this reason.  The webpack plugin can also generate the associate calls automatically(). If you want more info about using this, I'd be glad to assist.

Otherwise, you can call associate yourself.  

`CustomFunctions.associate("FOO.CALCINTEREST", fooCalcInterest);`

**Comment by wh1t3cAt1k** [comment]
Yes, I see that you can also do it using a JSDoc attribute:

```
/**
 * @customfunction MY.NAME MY.NAME
 */
 ...
```

**Comment by keyur32** [comment]
Marking this as closed as issue appears resolved.
